### PR TITLE
Create new post from defaults

### LINKED ISSUE #174: New post with default "settings"
On 18f.gsa.gov we require blog posts to have a title, authors, tags, and a few other things. It'd be handy if there were a button somewhere in the edit interface that adds a new document in a folder with any required settings added at post creation.

The UX for a site owner might be having an interface for adding required settings on a per-collection basis (or something).

The UX for the editor might be:
1. Click to add a new post
2. Fill in a webform with any required settings
3. Go right into editing the content

I'd be interested in helping out with this, I've done some work with validating frontmatter in Jekyll. This seems somewhat dependent on #166 (or at least lower priority).

### LINKED ISSUE #228: Add new page cannot be canceled
Right now, you can't exit out of the create a new page once you have clicked the button. Clicking cancel will still try and issue the `PUT` request. You can see this if you open your dev console and:

0. Navigate to one of your sites page listing
0. Click new page
0. Click cancel
0. See that the `PUT` request is still made, but the URL returns a 404 since it doesn't have a filename in the path.

If you are already editing a page, the `PUT` request will go throw and remove all of the content in file you are currently editing.

### PR DESCRIPTION
**Pull request by jeremiak**
Changes the add new page behavior to trigger an empty editor (with any post defaults as defined in `_config.yml`), and only saves that page back to Github when the user clicks save. The filename is derived from any `title` attribute in the front matter or a unique string in the case that there is no title.

If the user navigates away from the page without saving then the page won't be committed to Github (to close #228)

This should close #174, #224 and #228

### CODE REVIEW

File: assets/app/views/editor/edit-file.js [hunk 1]
```diff
@@ -143,13 +153,43 @@ var EditorView = Backbone.View.extend({
     if (settings) this.doc.frontMatter = settings;
     if (content) this.doc.content = content;
 
-    this.model.commit({
-      content: this.doc.toMarkdown(),
-      message: this.$('#save-content-message').val()
-    });
+    if (this.isNewPage) {
+      pageTitle = fileNameFromTitle(yaml.parse(settings)['title']);
+      pageTitle = pageTitle.toLowerCase().replace(/\s/g, '-');
```
Comment (dhcole) [inline assets/app/views/editor/edit-file.js]: Should we tighten this to something like `/^\w/`, which should filter for only alphanumeric and _ characters

Comment (jeremiak) [inline assets/app/views/editor/edit-file.js]: Ya, I think that's a good idea. Going to use `/\W/`

File: package.json [hunk 2]
```diff
@@ -59,7 +58,8 @@
     "to-markdown": "^1.2.1",
     "underscore": "^1.8.3",
     "uswds": "git+https://github.com/18F/web-design-standards.git#v0.9.x",
-    "validator": "^3.39.0"
+    "validator": "^3.39.0",
+    "yamljs": "^0.2.4"
```
Comment (dhcole) [inline package.json]: @jeremiak: Curious what made you switch libraries?

File: assets/app/models/Github.js [hunk 3]
```diff
@@ -173,16 +186,15 @@ var GithubModel = Backbone.Model.extend({
...
-      commitOpts = {
-        path: opts.path,
-        message: opts.message || 'The file ' + opts.path + ' was created',
-        content: opts.content || content
-      };
-    this.commit(commitOpts);
+  getDefaults: function () {
+    var config = this.configFiles['_config.yml'],
+        defaultConfigs = (config.present && config.json) ? config.json.defaults : [],
+        defaults = defaultConfigs.filter(function (c) {
```
Comment (dhcole) [inline assets/app/models/Github.js]: @jeremiak getting an error here if `defaultConfigs` is undefined, `.filter` is not available on `undefined`. What should happen if there is not a default section of the config file?

Comment (jeremiak) [inline assets/app/models/Github.js]: I think that if condition isn't complete. It should be something like `config.present && config.json && config.json.defaults` but that's pretty verbose. Essentially defaultConfigs should always at least be an empty array.

File: package.json (hunk 2 above)
Comment (jeremiak) [inline package.json]: There was a problem with browserifying `js-yaml` for some reason. We experienced the same thing in 18F/about-yml-npm#3 so I just switched to the other library here as well.

File: assets/app/models/Github.js [hunk 4]
```diff
@@ -116,9 +129,9 @@ var GithubModel = Backbone.Model.extend({
           url: url,
           complete: function(res) {
             var r = {
-              file: file,
+              file: file.name,
               present: (res.status === 200) ? true : false,
-              json: res.responseJSON || []
+              json: res.responseJSON || yaml.parse(res.responseText)
```
Comment (dhcole) [inline assets/app/models/Github.js]: This throws an error when one of the config / navigation files is not available:

```
GET http://localhost:1337/site/dhcole/microsite-template/_navigation.json 404 (Not Found)send @ bundle.js:24812jQuery.extend.ajax @ bundle.js:24348(anonymous function) @ bundle.js:246(anonymous function) @ bundle.js:2335(anonymous function) @ bundle.js:1885_arrayEach @ bundle.js:1737_each @ bundle.js:1728async.forEachOf.async.eachOf @ bundle.js:1884_parallel @ bundle.js:2334async.parallel @ bundle.js:2348Backbone.Model.extend.fetchConfig @ bundle.js:260Backbone.Model.extend.initialize @ bundle.js:145Backbone.Model @ bundle.js:3259child @ bundle.js:4703Backbone.View.extend.initialize @ bundle.js:823Backbone.View @ bundle.js:4043child @ bundle.js:4703Backbone.View.extend.edit @ bundle.js:1194Backbone.Router.extend.edit @ bundle.js:47_.extend.execute @ bundle.js:4336(anonymous function) @ bundle.js:4324(anonymous function) @ bundle.js:4607_.some._.any @ bundle.js:45924_.extend.loadUrl @ bundle.js:4605(anonymous function) @ bundle.js:28(anonymous function) @ bundle.js:3183(anonymous function) @ bundle.js:46556triggerEvents @ bundle.js:3225triggerApi @ bundle.js:3211eventsApi @ bundle.js:3009Events.trigger @ bundle.js:3201_.extend.set @ bundle.js:3381options.success @ bundle.js:3448fire @ bundle.js:19281self.fireWith @ bundle.js:19393done @ bundle.js:24446(anonymous function) @ bundle.js:24787
bundle.js:47951 Uncaught <ParseException> You cannot define a sequence item when in a mapping
``

**Comment by dhcole** [comment]
Getting an error when saving a new post without a `title`

```
Uncaught TypeError: Cannot read property 'title' of null
Backbone.View.extend.saveDocument @ bundle.js:752
jQuery.event.dispatch @ bundle.js:20617
elemData.handle @ bundle.js:20303
```

**Comment by dhcole** [comment]
Works well now. :ship:

**Pull request merged**